### ISSUE TITLE
proposal: `annotate_redeclares`

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#4747]
## Description

Annotate redeclared members.

## Details

**DO** annotate redeclared members.

This practice improves code readability and helps protect against unintentionally redeclaring members or being surprised when a member ceases to redeclare (due for example to a rename refactoring).

## Kind

Style  and errors.

## Bad Examples

```dart
class C { 
  void f() { }
}

extension type E(C c) implements C {
  void f() {
    ...
  }
}
```

## Good Examples 

```dart
import 'package:meta/meta.dart';

class C { 
  void f() { }
}

extension type E(C c) implements C {
  @redeclare
  void f() {
    ...
  }
}
```

## Discussion

This is essentially an analog to `annotate_overrides`.

### Discussion checklist

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [x] If there's any prior art (e.g., in other linters), please add references here.
- [ ] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [ ] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by pq** [comment]
@srawlins @bwilkerson @stereotype441: it feels like we have a bit of a consensus on the value of this one. If that's right I'll bump this to accepted and implement.

**Comment by srawlins** [comment]
I support!

**Comment by pq** [comment]
Bumping to accepted.

**Comment by grouma** [comment]
+acx-web-core-eng ***@***.***>

**Comment by pq** [comment]
Implemented w/ https://github.com/dart-lang/sdk/commit/d1b6ea9f4d1514fa8332af6523e9eca0597f44e5.
